spec.py

```diff
@@ -0,0 +1,297 @@
+
+  @abc.abstractproperty
+  def max_allowed_runtime(self):
+    """The max allowed runtime of the workload."""
+
+  @abc.abstractproperty
+  def eval_period_time(self):
+    """The eval period of the workload."""
+
+  @abc.abstractmethod
+  def is_output_params(self, param_key: ParameterKey) -> bool:
+    """Whether or not a key in ParameterTree is the output layer parameters."""
+
+  @abc.abstractmethod
+  def preprocess_for_train(
+      self,
+      selected_raw_input_batch: Tensor,
+      selected_label_batch: Tensor,
+      train_mean: Tensor,
+      train_stddev: Tensor,
+      rng: RandomState) -> Tensor:
+    """return augmented_and_preprocessed_input_batch"""
+
+  @abc.abstractmethod
+  def preprocess_for_eval(
+      self,
+      raw_input_batch: Tensor,
+      train_mean: Tensor,
+      train_stddev: Tensor) -> Tensor:
+    """return preprocessed_input_batch"""
+
+  # InitModelFn = Callable[
+  #     Tuple[ParameterShapeTree, RandomState], ParameterTree]
+  @abc.abstractmethod
+  def init_model_fn(
+      self,
+      param_shapes: ParameterShapeTree,
+      rng: RandomState) -> [ParameterTree, ModelAuxillaryState]:
+    """return initial_params"""
+
+  # ModelFn = Callable[
+  #     Tuple[ParameterTree, Tensor, ForwardPassMode, RandomState, bool],
+  #     Tensor]
+  @abc.abstractmethod
+  def model_fn(
+      self,
+      params: ParameterTree,
+      augmented_and_preprocessed_input_batch: Tensor,
+      model_state: ModelAuxillaryState,
+      mode: ForwardPassMode,
+      rng: RandomState,
+      update_batch_norm: bool) -> Tuple[Tensor, ModelAuxillaryState]:
+    """return logits_batch"""
+    # Possible side effect of updating BN.
+
+  # Keep this separate from the loss function in order to support optimizers
+  # that use the logits.
+  def output_activation_fn(
+      self,
+      logits_batch: Tensor,
+      loss_type: LossType) -> Tensor:
+    if loss_type == LossType.SOFTMAX_CROSS_ENTROPY:
+      return jax.nn.softmax(logits_batch, axis=-1)
+    if loss_type == LossType.SIGMOID_CROSS_ENTROPY:
+      return jax.nn.sigmoid(logits_batch)
+    if loss_type == LossType.MEAN_SQUARED_ERROR:
+      return logits_batch
+
+  # LossFn = Callable[Tuple[Tensor, Tensor], Tensor]
+  # Does NOT apply regularization, which is left to the submitter to do in
+  # `update_params`.
+  @abc.abstractmethod
+  def loss_fn(
+      self,
+      label_batch: Tensor,
+      logits_batch: Tensor,
+      loss_type: LossType) -> Tensor:  # differentiable
+    """return oned_array_of_losses_per_example"""
+
+  @abc.abstractmethod
+  def eval_model(
+      self,
+      params: ParameterTree,
+      model_state: ModelAuxillaryState,
+      rng: RandomState):
+    """Run a full evaluation of the model."""
+
+
+class TrainingCompleteError(Exception):
+  pass
+
+
+# Training algorithm track submission functions, to be filled in by the
+# submitter.
+
+
+def init_optimizer_state(
+    workload: Workload,
+    param_shapes: ParameterShapeTree,
```

Comment:
Do we need `param_shapes` as an argument given we already take `workload`?

Reply:
Removed.